One-dimensionalization of data models

LzLang opened the pull request:
Dear @BrapiCoordinatorSelby,

@VivianBass and I finished the editing of the schemas. Like we talked about, we added the `relationshipType` to every occurence of the attributes `additionalInfo` and `externalReferences` in every model. Also we flattend all models, so there shouldn`t be a nested model left or rather all models should be 1-dimensional. (As in Wittenberg, we proceeded here using separate models and associations).

In some models there were location specifications (such as ImageLocation in Image.json). These attributes were effectively just a copy of GeoJSON, although GeoJSON exists as an independent model. Accordingly, we have changed the attributes so that they form an association to GeoJSON: GeoJSON
```json
"GeoJSON": {
	"title": "GeoJSON",
	"type": "object",
	"description": "One geometry as defined by GeoJSON (RFC 7946). All coordinates are decimal values on the WGS84 geographic coordinate reference system.\n\nCopied from RFC 7946 Section 3.1.1\n\nA position is an array of numbers. There MUST be two or more elements. The first two elements are longitude and latitude, or\neasting and northing, precisely in that order and using decimal numbers. Altitude or elevation MAY be included as an optional third element.",
	"properties": {
		"geometryDbId": {
			"description": "Unique identifier for the geometry",
			"type": [
				"null",
				"string"
			]
		},
		"geometry": {
			"description": "A geometry as defined by GeoJSON (RFC 7946). In this context, only Point or Polygon geometry are allowed.",
			"$ref": "GeoJSONGeometry.json#/$defs/GeoJSONGeometry"
		},
		"type": {
			"type": "string",
			"default": "Feature",
			"example": "Feature",
			"description": "The literal string \"Feature\""
		},
		"image": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Phenotyping/Image.json#/$defs/Image",
			"relationshipType": "many-to-one",
			"referencedAttribute": "imageLocation"
		},
		"observation": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Phenotyping/Observation.json#/$defs/Observation",
			"relationshipType": "many-to-one",
			"referencedAttribute": "geoCoordinates"
		},
		"observationUnit": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Phenotyping/ObservationUnit.json#/$defs/ObservationUnitPosition",
			"relationshipType": "many-to-one",
			"referencedAttribute": "geoCoordinates"
		},
		"germplasmOrigin": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Germplasm/Germplasm.json#/$defs/GermplasmOrigin",
			"relationshipType": "many-to-one",
			"referencedAttribute": "coordinates"
		}
	},
	"required": [
		"geometryDbId"
	]
}
```

New `ImageLocation` from `Image.json`:
```json
"imageLocation": {
	"description": "One geometry as defined by GeoJSON (RFC 7946). All coordinates are decimal values on the WGS84 geographic coordinate reference system.\n\nCopied from RFC 7946 Section 3.1.1\n\nA position is an array of numbers. There MUST be two or more elements. The first two elements are longitude and latitude, or\neasting and northing, precisely in that order and using decimal numbers. Altitude or elevation MAY be included as an optional third element.",
	"relationshipType": "one-to-many",
	"referencedAttribute": "image",
	"items": {
		"$ref": "../BrAPI-Common/GeoJSON.json#/$defs/GeoJSON",
		"description": "A geometry as defined by GeoJSON (RFC 7946). In this context, only Point or Polygon geometry are allowed."
	},
	"title": "GeoJSON",
	"type": [
		"null",
		"array"
	]
}
```
Could you please check or changes, if there is something wrong or if you like us to change other things?

Thank you and best regards! :)

LzLang committed:
One-dimensionalization of data models

Dear @BrapiCoordinatorSelby,

@VivianBass and I finished the editing of the schemas.
Like we talked about, we added the `relationshipType` to every occurence of the attributes `additionalInfo` and `externalReferences` in every model.
Also we flattend all models, so there shouldn`t be a nested model left or rather all models should be 1-dimensional.
(As in Wittenberg, we proceeded here using separate models and associations).

In some models there were location specifications (such as ImageLocation in Image.json). These attributes were effectively just a copy of GeoJSON, although GeoJSON exists as an independent model.
Accordingly, we have changed the attributes so that they form an association to GeoJSON:
GeoJSON
```json
"GeoJSON": {
	"title": "GeoJSON",
	"type": "object",
	"description": "One geometry as defined by GeoJSON (RFC 7946). All coordinates are decimal values on the WGS84 geographic coordinate reference system.\n\nCopied from RFC 7946 Section 3.1.1\n\nA position is an array of numbers. There MUST be two or more elements. The first two elements are longitude and latitude, or\neasting and northing, precisely in that order and using decimal numbers. Altitude or elevation MAY be included as an optional third element.",
	"properties": {
		"geometryDbId": {
			"description": "Unique identifier for the geometry",
			"type": [
				"null",
				"string"
			]
		},
		"geometry": {
			"description": "A geometry as defined by GeoJSON (RFC 7946). In this context, only Point or Polygon geometry are allowed.",
			"$ref": "GeoJSONGeometry.json#/$defs/GeoJSONGeometry"
		},
		"type": {
			"type": "string",
			"default": "Feature",
			"example": "Feature",
			"description": "The literal string \"Feature\""
		},
		"image": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Phenotyping/Image.json#/$defs/Image",
			"relationshipType": "many-to-one",
			"referencedAttribute": "imageLocation"
		},
		"observation": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Phenotyping/Observation.json#/$defs/Observation",
			"relationshipType": "many-to-one",
			"referencedAttribute": "geoCoordinates"
		},
		"observationUnit": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Phenotyping/ObservationUnit.json#/$defs/ObservationUnitPosition",
			"relationshipType": "many-to-one",
			"referencedAttribute": "geoCoordinates"
		},
		"germplasmOrigin": {
			"description": "Geometry associated with an image",
			"$ref": "../BrAPI-Germplasm/Germplasm.json#/$defs/GermplasmOrigin",
			"relationshipType": "many-to-one",
			"referencedAttribute": "coordinates"
		}
	},
	"required": [
		"geometryDbId"
	]
}
```

New `ImageLocation` from `Image.json`:
```json
"imageLocation": {
	"description": "One geometry as defined by GeoJSON (RFC 7946). All coordinates are decimal values on the WGS84 geographic coordinate reference system.\n\nCopied from RFC 7946 Section 3.1.1\n\nA position is an array of numbers. There MUST be two or more elements. The first two elements are longitude and latitude, or\neasting and northing, precisely in that order and using decimal numbers. Altitude or elevation MAY be included as an optional third element.",
	"relationshipType": "one-to-many",
	"referencedAttribute": "image",
	"items": {
		"$ref": "../BrAPI-Common/GeoJSON.json#/$defs/GeoJSON",
		"description": "A geometry as defined by GeoJSON (RFC 7946). In this context, only Point or Polygon geometry are allowed."
	},
	"title": "GeoJSON",
	"type": [
		"null",
		"array"
	]
}
```
Could you please check or changes, if there is something wrong or if you like us to change other things?

Thank you and best regards! :)

daveneti commented:
Thanks @LzLang, I will check it out and run it through the schema validator. 

I have a suggestion about the additions (not the re-organization part). Would if be possible to put this in a different branch, since these change the model? So keep the data-model-separation for v2.1? Alternatively we could merge the data model in to 2.1 branch and then create a new branch for 2.2 model changes?

What do you think @BrapiCoordinatorSelby?

daveneti commented:
@LzLang I got the following validation errors:
There were 5 errors validating the JSON Schema:
Validation Error :	The required property 'progenyDbId' is not found in the list of properties of 'Progeny', expecting one of 'parentDbId, progenyGermplasm, parentType, pedigreeNode'
Validation Error :	The required property 'description' is not found in the list of properties of 'EnvironmentParameters', expecting one of 'environmentParametersDbId, parameterName, parameterPUI, unit, unitPUI, value, valuePUI, study'
Validation Error :	The required property 'crossDbId' is not found in the list of properties of 'CrossAttributes', expecting one of 'attributeDbId, crossAttributeName, crossAttributeValue, cross'
Validation Error :	The required property 'storageDbId' is not found in the list of properties of 'StorageTypes', expecting one of 'storageDbID, code, description, germplasm'
Validation Error :	The required property 'originDbId' is not found in the list of properties of 'GermplasmOrigin', expecting one of 'originDbID, coordinateUncertainty, coordinates, germplasm'

I will make the corrections and commit

Commit:
reformatted to 4 space indent

daveneti commented on Specification/BrAPI-Schema/BrAPI-Common/Species.json:
```diff
@@ -0,0 +1,51 @@
+    "$defs": {
+        "Species": {
+			"properties": {
+				"specieDbId": {
```
I think this should be speciesDbId

daveneti commented on Specification/BrAPI-Schema/BrAPI-Common/Variable.json:
```diff
@@ -4,7 +4,15 @@
             "properties": {
                 "additionalInfo": {
                     "description": "A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestricted by the BrAPI specification.",
-                    "$ref": "../BrAPI-Common/AdditionalInfo.json#/$defs/AdditionalInfo"
```
This changes the model, from a single to an array of AdditionalInfo, can this be put in v2.2? See in same changes in other entities.

daveneti commented on Specification/BrAPI-Schema/BrAPI-Core/Study.json:
```diff
@@ -66,90 +74,38 @@
+                    "items": {
+                        "$ref": "../BrAPI-Core/Study.json#/$defs/ExperimentalDesign",
+                        "description": "ExperimentalDesign"
                     },
```
Model change from a single ExperimentalDesign to an array of ExperimentalDesign. Not sure this is appropriate.

daveneti commented on Specification/BrAPI-Schema/BrAPI-Core/Study.json:
```diff
@@ -159,47 +115,30 @@
                     "type": [
                         "null",
-                        "object"
+                        "array"
```
Model change from a single grow facility to array of growth facilities!

daveneti commented on Specification/BrAPI-Schema/BrAPI-Core/Study.json:
```diff
@@ -159,47 +115,30 @@
                     "type": [
                         "null",
-                        "object"
+                        "array"
```
Model change from single lastUpdate to array of lastUpdates!

daveneti commented on Specification/BrAPI-Schema/BrAPI-Genotyping/Reference.json:
```diff
@@ -111,26 +109,16 @@
                     "type": [
                         "null",
-                        "object"
+                        "array"
```
change from single to array, model change!

daveneti commented on Specification/BrAPI-Schema/BrAPI-Genotyping/ReferenceSet.json:
```diff
@@ -105,25 +103,16 @@
+                    "title": "Species",
                     "type": [
                         "null",
-                        "object"
```
Model change from single to array.

daveneti commented on Specification/BrAPI-Schema/BrAPI-Genotyping/VariantSet.json:
```diff
@@ -50,146 +58,27 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Genotyping/VariantSet.json#/$defs/AvailableFormats",
```
I think this should be singular 'AvailableFormat'

daveneti commented on Specification/BrAPI-Schema/BrAPI-Genotyping/VariantSet.json:
```diff
@@ -215,40 +106,13 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Genotyping/VariantSet.json#/$defs/MetadataFields",
```
I think this should be singular 'MetadataField'

daveneti commented on Specification/BrAPI-Schema/BrAPI-Genotyping/VariantSet.json:
```diff
@@ -307,6 +171,222 @@
+                "primaryModel": false
+            }
+        },
+        "AvailableFormats": {
```
Should be AvaliableFormat

daveneti commented on Specification/BrAPI-Schema/BrAPI-Genotyping/VariantSet.json:
```diff
@@ -307,6 +171,222 @@
+                "primaryModel": false
+            }
+        },
+        "MetadataFields": {
```
Should be MetadataField

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -4,30 +4,25 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Germplasm/Cross.json#/$defs/CrossAttributes",
```
Should be CrossAttribute

LzLang replied:
Should I change the relationshipType to one-to-one too or is one-to-many preferred becaus it's "crossAttributes"?

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -81,33 +78,13 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Germplasm/Cross.json#/$defs/PollinationEvents",
```
Should be PollinationEvent

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -136,6 +113,86 @@
                 "primaryModel": true
             }
+        },
+        "CrossAttributes": {
```
Should be CrossAttribute

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -136,6 +113,86 @@
+                "primaryModel": true
+            }
+        },
+        "PollinationEvents": {
```
Should be PollinationEvent

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -105,34 +113,24 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Germplasm/Germplasm.json#/$defs/Donors",
```
Should be Donor

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -259,37 +229,13 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Germplasm/Germplasm.json#/$defs/StorageTypes",
```
Should be StorageType

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -311,49 +257,27 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Germplasm/Germplasm.json#/$defs/Synonyms",
```
should be Synonym or better still GermplasmSynonym

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -466,6 +390,201 @@
                 "primaryModel": true
             }
+        },
+        "Donors": {
```
Should be Donor

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -466,6 +390,201 @@
+                "primaryModel": false
+            }
+        },
+        "StorageTypes": {
```
Should be StorageType

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -466,6 +390,201 @@
+                "primaryModel": false
+            }
+        },
+        "TaxonIds": {
```
Should be TaxonId

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -58,26 +68,13 @@
-                            "parentType"
-                        ],
-                        "type": "object"
+                        "$ref": "../BrAPI-Germplasm/PedigreeNode.json#/$defs/Parents",
```
Should be GermplasmParent

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -139,6 +123,74 @@
                 "primaryModel": true
             }
+        },
+        "Parents": {
```
Should be GermplasmParent

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Event.json:
```diff
@@ -55,63 +41,13 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Phenotyping/Event.json#/$defs/EventParameters",
```
Should be EventParameter

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Image.json:
```diff
@@ -68,24 +78,16 @@
                     "type": [
                         "null",
-                        "object"
+                        "array"
```
Model change from single to array

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Observation.json:
```diff
@@ -26,24 +36,16 @@
                     "type": [
                         "null",
-                        "object"
+                        "array"
```
Model change from single to array

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/OntologyReference.json:
```diff
@@ -4,30 +4,13 @@
-                            }
-                        },
-                        "type": "object"
+                        "$ref": "../BrAPI-Phenotyping/OntologyReference.json#/$defs/DocumentationLinks",
```
Should be DocumentationLink

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -68,51 +78,16 @@
+                    "relationshipType": "one-to-many",
+                    "referencedAttribute": "scale",
+                    "items": {
+                        "$ref": "../BrAPI-Phenotyping/Scale.json#/$defs/ValidValues",
```
Should be ValidValue

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -125,6 +100,92 @@
                 "primaryModel": true
             }
+        },
+        "ValidValues": {
```
Should be ValidValue

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -125,6 +100,92 @@
+                "primaryModel": false
+            }
+        },
+        "Categories": {
```
Should be Category

daveneti reviewed (changes requested):
I notice a few model changes from single to array and use of Plural when Singular is appropriate.

Commit:
added back missing 'description' property

Commit:
corrected case and spelling

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -92,26 +89,13 @@
-                            "parentType"
-                        ],
-                        "type": "object"
+                        "$ref": "../BrAPI-Germplasm/PedigreeNode.json#/$defs/Progeny",
```
I prefer the team GermplasmChild, since it is more specific.

LzLang replied:
Would you prefere, that we rename the whole "Progeny" Model to "GermplasmChild"?

daveneti replied:
Yes please change :)

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -136,7 +113,87 @@
+				}
+			},
+			"required": [
+                "crossDbId"
```
Listed as required but is not a property of CrosssAttribute

LzLang replied:
Changed it accordingly to `crossAttributeDbId`

daveneti reviewed (changes requested):
Added some more comments to previous review.

LzLang commented:
Thank you very much @daveneti for your review and corrections!
I didn't expect that we would have introduced so many new mistakes without realizing it.
Therefore thank you so much for correcting those!

> I have a suggestion about the additions (not the re-organization part). Would if be possible to put this in a different branch, since these change the model? So keep the data-model-separation for v2.1? Alternatively we could merge the data model in to 2.1 branch and then create a new branch for 2.2 model changes?

Personally I don't see a problem with putting it in a different branch instaed of merching it.
My current status was that the goal is to generally one-dimensionalize the BrAPI models so that there are no more nested attributes/models.
However, I don't know what influence this change will have on other projects.
Nonetheless, I would like to sit down with Mr. Hallab again next week to discuss this and get his opinion and view on thise matter.



I have now noticed something else:
In some models, for example `AdditionalInfo` and `TraitDataType`, the “properties” have been removed.
```json
{
    "$defs": {
        "TraitDataType": {
            "description": "<p>Class of the scale, entries can be</p>\n<p>\"Code\" -  This scale class is exceptionally used to express complex traits. Code is a nominal scale that combines the expressions of the different traits composing the complex trait. For example a severity trait might be expressed by a 2 digit and 2 character code. The first 2 digits are the percentage of the plant covered by a fungus and the 2 characters refer to the delay in development, e.g. \"75VD\" means \"75 %\" of the plant is infected and the plant is very delayed.</p>\n<p>\"Date\" - The date class is for events expressed in a time format, See ISO 8601</p>\n<p>\"Duration\" - The Duration class is for time elapsed between two events expressed in a time format, e.g. days, hours, months</p>\n<p>\"Nominal\" - Categorical scale that can take one of a limited and fixed number of categories. There is no intrinsic ordering to the categories</p>\n<p>\"Numerical\" - Numerical scales express the trait with real numbers. The numerical scale defines the unit e.g. centimeter, ton per hectare, branches</p>\n<p>\"Ordinal\" - Ordinal scales are scales composed of ordered categories</p>\n<p>\"Text\" - A free text is used to express the trait.</p>",
            "enum": [
                "Code",
                "Date",
                "Duration",
                "Nominal",
                "Numerical",
                "Ordinal",
                "Text"
            ],
            "type": "string",
            "example": "Numerical"
        }
    },
    "$id": "https://brapi.org/Specification/BrAPI-Schema/Components/Common/AdditionalInfo.json",
    "$schema": "http://json-schema.org/draft/2020-12/schema"
}
```
Is there a specific reason why these have been removed?
Personally, I think it would make sense to add them back for the sake of consistency.



Once again, thank you very much for your review.
I'll look over you review in detail and talk with @VivianBass about it :)

daveneti commented:
Hi @LzLang 

Thanks for your comment and changes. Yes, creating a new branch for v2.2 is a great idea, perhaps that can be done with the other 2.2 changes @BrapiCoordinatorSelby will add. I am not actually sure if there any many model changes. I think I might have been confused. If you can review my suggestions and make any changes, then we can review again to be sure. The comment you made about properties I will double check when get back to my computer.

daveneti commented:
Hi @LzLang, back at my computer now. 

For AdditionalInfo I did not see any properties on the original definition, only addtionalProperties, which is still there.
TraitDataType, is an enum, which do not have any properties.

Can you explain further your comment on theses please? 

I can see in the example the id is incorrect. That looks Iike C&P error, I will correct that. It should be     "$id": "https://brapi.org/Specification/BrAPI-Schema/Components/Common/TraitDataType.json"

I will add a check in the validator that looks for those mistakes.

Commit:
corrected TraitDataType id

LzLang commented:
Hey @daveneti,

> For AdditionalInfo I did not see any properties on the original definition, only addtionalProperties, which is still there. TraitDataType, is an enum, which do not have any properties.

Sorry, for the confusion! This is a mistake on my part.
When we did the model separation in Wittenberg we created AditionalInfo as separated model.
In our version AdditionalInfo included the `properties` key, that cotained attributes like reverse association.
Due a fault on my side I never committed our changes officially, therefore there are no `properties` in the original model (when I wrote my comment, I forgot this and mixed it up).
In the current converter version, we work around this by adding the reverse associations in AdditionalInfo on the code side. Therefore we also needed the `relationshipType` in the respective models that had an association to AdditionalInfo.
For example, the association of `Study.json`:
```json
"additionalInfo": {
	"description": "A free space containing any additional information related to a particular object. A data source may provide any JSON object, unrestricted by the BrAPI specification.",
	"$ref": "../BrAPI-Common/AdditionalInfo.json#/$defs/AdditionalInfo"
	"relationshipType": "one-to-many",
	"items": {
		"$ref": "../BrAPI-Common/AdditionalInfo.json#/$defs/AdditionalInfo",
		"description": "AdditionalInfo"
	},
	"type": [
		"null",
		"array"
	]
}
```


If @asishallab and @VivianBass  have no objections, I think we can adopt the changes without further problems and use them as a basis for v2.2 and our own project.

daveneti commented:
Thanks @LzLang 

In the above 'additionalInfo' example, becareful you have the ref to two places AdditionalInfo. I don't think it should be in the 'items', and the 'type' should be removed.  

Can you make all the changes in this branch, then I can review again, before merging into 'data-model-separation'. I will need to check again with the validator and then I can merge this modified 2.1 version into our extended versions.  I don't think there are any model changes so this would be still be 2.1.

LzLang commented:
> In the above 'additionalInfo' example, be careful you have the $ref to two places. I don't think it should be in the 'items', and the 'type' should be removed. Also I think the relationship will be one-to-one. E.g.

Thank you @daveneti for your advice, this simple slipped my eyes.
I will check the models again and see if this mistake occured just once or if I made this mistake in other models too.
While testing I noticed that I only added the `relationshipType` to `additionalInfo` and `externalReferences`, but there are other associations that don't have a relationshipType, therefore I would like to add those too.



> Can you make all the changes in this branch, then I can review again, before merging into 'data-model-separation'. 

Certainly!



> I will need to check again with the validator and then I can merge this modified 2.1 version into our extended versions. I don't think there are any model changes so this would be still be 2.1.

There are some model changes, because there were some nested models left, which we one-dimensionalized.
However I can give you a list of the changed/new models, so we can decide what is the best course of action for merging.

LzLang commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -139,6 +123,74 @@
+        },
+        "Progeny": {
+            "properties": {
+                "parentDbId": {
```
Changed it to progenyDbId
Zendro is using the models for a data warehouse.
The problem we have is, that we need a primary key, which is why I added it.
If @asishallab don't mind I could remove it and add the primary key code side.

LzLang commented:
So far I changed the models accordingly to our conversation.
Also I added one-to-one relationship type to associations that were singular and missing a type.

For me, there are more problems with plural associations.
`Study.json`:
```json
"dataLinks": {
	"description": "List of links to extra data files associated with this study. Extra data could include notes, images, and reference data.",
	"items": {
		"$ref": "DataLink.json#/$defs/DataLink"
	},
	"type": [
		"null",
		"array"
	]
}
```

`Cross.json`:
```json
"parent1": {
	"$ref": "CrossParent.json#/$defs/CrossParent",
	"description": "the first parent used in the cross"
},
"parent2": {
	"$ref": "CrossParent.json#/$defs/CrossParent",
	"description": "the second parent used in the cross"
}
```

`CrossingProject.json`:
```json
"potentialParents": {
	"description": "A list of all the potential parents in the crossing block, available in the crossing project\n<br/> If the parameter 'includePotentialParents' is false, the array 'potentialParents' should be empty, null, or excluded from the response object.",
	"items": {
		"$ref": "CrossParent.json#/$defs/CrossParent"
	},
	"type": [
		"null",
		"array"
	]
}
```

`PlannedCross.json`:
```json
"parent1": {
	"$ref": "CrossParent.json#/$defs/CrossParent",
	"description": "the first parent used in the cross"
},
"parent2": {
	"$ref": "CrossParent.json#/$defs/CrossParent",
	"description": "the second parent used in the cross"
}
```
What relationshipType would you prefer for those associations?


Another thing I noticed while revising:
There isn't an uniform structure for 'one-to-many' and 'many-to-many' associations.
`PlannedCross.json`:
The informations are inside and outside of `items`
```json
"crosses": {
	"title": "crosses",
	"description": "crosses",
	"referencedAttribute": "plannedCross",
	"relationshipType": "one-to-many",
	"items": {
		"$ref": "Cross.json#/$defs/Cross",
		"description": "Cross"
	},
	"type": [
		"null",
		"array"
	]
}
```

`PedigreeNode.json`:
The informations are all in `items`
```json
"siblings": {
	"description": "A list of sibling germplasm references in the pedigree tree for this germplasm. These represent edges in the tree, connecting to other nodes.\n<br/> Siblings share at least one parent with the given germplasm. \n<br/> If the parameter 'includeSiblings' is set to false, then this array should be empty, null, or not present in the response.",
	"items": {
		"$ref": "Germplasm.json#/$defs/Germplasm",
		"description": "The ID which uniquely identifies a parent germplasm",
		"referencedAttribute": "siblingPedigreeNodes",
		"relationshipType": "many-to-one"
	},
	"type": [
		"null",
		"array"
	]
}
```

The first definition seems like the old one.
So would you prefere, that we change plural associations to the new definition?

daveneti commented:
@LzLang Thanks for all your work, there is a lot to unpack there and consider.

On the question of which relationship type:
Study < datalinks > DataLink - The DataLink is not an entity, but a Value Type. It can no exist outside of a specific Study. So the relationship will be 'many-to-one'. Many data links associated with a single study
Cross < parent1 & parent2 > CrossParent - This is a straight 'one-to-one' relationship. BTW the schema validator assumes a 'one-to-one' relationship if the property references a single object or 'many-to-one' if the property references an array.
CrossingProject < potentialParents > CrossParent. A CrossParent is a value type, so although the relationship between a CrossingProject and it's potential parents is 'many-to-many', the relationship between CrossingProject and CrossParent is 'many-to-one'. BTW the relationship between CrossParent and Germplasm or ObservationUnit will be 'one-to-many'. So you could say 'many-to-one' + 'one-to-many' = 'many-to-many'
PlannedCross < parent1 & parent2 > CrossParent - This is the same as Cross < parent1 & parent2 > 'one-to-one' 

You should add the relationship between CrossParent and Germplasm or ObservationUnit as 'one-to-many

Well spotted on the non-uniform placement of the referencedAttribute and relationshipType. I had not noticed these! I propose that these are associated with the property and not the items in a array. So the example for crosses in PlannedCross. The schema validator will only look for these in the property and not the items.  

I will now review the PR.

daveneti commented:
@LzLang Did you commit your changes? I did not see a new commit from you. Can you commit the changes to the branch and then I will review again.

If there are any model changes, we can put these in a separate branch. Putting the models in a separate top level definition if it does not change the cardinality of the relationship, is not a model change, so I think there are actually none or few model changes. Are there any additional properties or changes from one object to array or vice versa? I don't think there are any. I can do a final check on the PR once you have committed the changes.

LzLang committed:
Changed the models accordingly to the review

LzLang commented:
@daveneti I made the changes we talked about and commited it-

> Are there any additional properties or changes from one object to array or vice versa? I don't think there are any.

No, I don't think that we changed models in that way, or rather I reversed those changes according to your review.

So far I have made the changes we discussed.
At the moment I have refrained from adding the relationshipType, which we wrote about yesterday/today (but I will do so).
I also plan to go through the entire models again to make sure I haven't missed anything that could lead to errors in general or with Zendro (including minor errors or inconsistent definitions (e.g. the case with items)).
However, it would certainly be advantageous to create a separate PR here.

I will of course fix the errors/problems you find in the review as soon as possible.

We still have one question regarding [GeoJSONGeometry](https://github.com/plantbreeding/BrAPI/blob/data-model-separation/Specification/BrAPI-Schema/BrAPI-Common/GeoJSONGeometry.json):
This model uses “oneOf”, which Zendro does not natively support.
I would like to solve this problem on the code side (instead of changing anything in the model) by setting one of the two geometries as default with the possibility to optionally use the other geometry.
Which of the two geometries would you prefer as the “default”?

Sidenote: I'm currentyl helping my parents to move and I'm moving myself too, which is why my work on the models has been delayed somewhat.

daveneti commented:
@LzLang Thanks for the changes, I will review now

For GeoJSONGeometry I don't have a suggestion. Perhaps @BrapiCoordinatorSelby has one? In the Schema tools generator for GraphQL I added a TypeResolver, that simply returned the first type to get the schema to generate. Ideally, we need to use a discriminator, which is the 'type'property. Since I was not generating a working API, it did not matter to me, which was default. 

    @AllArgsConstructor
    private static class UnionTypeResolver implements TypeResolver {

        private GraphQLUnionType unionType;

        @Override
        public GraphQLObjectType getType(TypeResolutionEnvironment schemaName) {
            return (GraphQLObjectType) this.unionType.getTypes().get(0);
        }
    }

daveneti commented:
PS, @LzLang it would be great soon to compare the GraphQL schema you have developed and the one the Schema tools generates. I will add soon a CI build to generate it each time the model changes.

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -114,9 +107,9 @@
-        "CrossAttributes": {
+        "CrossAttribute": {
             "properties": {
-                "attributeDbId": {
```
No DbId in CrossAttribute in v2.1. Perhaps there should be @BrapiCoordinatorSelby, something to consider for v2.2

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -142,18 +135,18 @@
             },
             "required": [
-                "crossDbId"
+                "crossAttributeDbId"
```
See comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -142,18 +135,18 @@
-        "PollinationEvents": {
+        "PollinationEvent": {
             "properties": {
-                "eventDbId": {
```
No DbId in v2.1 for pollination event.

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Cross.json:
```diff
@@ -186,7 +179,7 @@
             },
             "required": [
-                "eventDbId"
+                "pollinationEventDbId"
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -462,17 +455,17 @@
                 }
```
In v2.1 coordinates is a single object not an array. I think this is a mistake in the original schema @BrapiCoordinatorSelby ?

LzLang replied:
Should I change/adjust anything here?
At least in the github overview, I don't see the correct place to which your comment refers.

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -462,17 +455,17 @@
             },
             "required": [
-                "originDbId"
+                "germplasmOriginDbId"
```
No DbId in germplasm original

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -510,15 +503,15 @@
             },
             "required": [
-                "storageDbId"
+                "storageTypeDbId"
```
storage type has no DbId in v2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/Germplasm.json:
```diff
@@ -554,9 +547,9 @@
+        "TaxonId": {
             "properties": {
-                "taxonDbId": {
+                "taxonIdDbId": {
```
Should remain as taxonDbId, this is how it is in v2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/GermplasmAttributeValue.json:
```diff
@@ -4,23 +4,16 @@
                     "referencedAttribute": "attributeValues",
                     "relationshipType": "many-to-one"
                 },
-                "attributeValueDbId": {
```
Should remain as attributeValueDbId, this is how it is in v2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -124,9 +124,9 @@
-        "Parents": {
+        "GermplasmParent": {
             "properties": {
-                "parentDbId": {
```
This is interesting issue. @BrapiCoordinatorSelby parentDbId was added to the JSON schema but is not part of v2.1. It was added to the JSON schema as 'Parents', with the additional properties parentDbId (replacing germplasmId) and pedigreeNode. To be totally consistent with other links there should a germplasm property with a link to a germplasm. Perhaps the pedigreeNode can be added in v2.2? Also renaming 'Parents' to 'GermplasmParent' is good idea and does not affect the model, but it is clearly because the object is a single parent.

LzLang replied:
I have removed the DbId again.
If I understand your comment correctly, it would otherwise be a model change again at the moment, wouldn't it?

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -158,9 +158,9 @@
             }
         },
-        "Progeny": {
+        "GermplasmChild": {
```
@BrapiCoordinatorSelby this was proposed because 'Progeny' is plural.

LzLang replied:
Should I change anything or leave it for now?

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -158,9 +158,9 @@
+        "GermplasmChild": {
             "properties": {
-                "parentDbId": {
+                "germplasmChildDbId": {
```
Same as my comment above on GermplasmParent

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -182,7 +182,7 @@
             },
             "required": [
-                "progenyDbId",
+                "germplasmChildDbId",
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/PedigreeNode.json:
```diff
@@ -158,9 +158,9 @@
-        "Progeny": {
+        "GermplasmChild": {
             "properties": {
-                "parentDbId": {
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/SeedLot.json:
```diff
@@ -167,7 +160,7 @@
+                "contentMixtureDbId": {
                     "description": "The unique DbId for a cross contained in this seed lot",
                     "type": [
                         "null",
```
To be consistent with other parts of the JSON Schema crossDbId and crossName should be replaced by a single property cross, with a link to cross. Similarly for germplasmDbId and germplasmName, with germplasm property and link to Germplasm

LzLang replied:
So basically an association to the referenced model, right?
>single property cross

So one-to-one relationship?

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/SeedLot.json:
```diff
@@ -167,7 +160,7 @@
         },
         "ContentMixture": {
             "properties": {
-                "mixtureDbId": {
```
No mixtureDbId in 2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/SeedLot.json:
```diff
@@ -210,7 +203,7 @@
                 }
             },
             "required": [
-                "mixtureDbId"
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/SeedLotTransaction.json:
```diff
@@ -46,7 +39,7 @@
                     "relationshipType": "many-to-one"
                 },
-                "transactionDbId": {
+                "seedLotTransactionDbId": {
```
it is transactionDbId in v2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Germplasm/SeedLotTransaction.json:
```diff
@@ -74,7 +67,7 @@
             },
             "required": [
-                "transactionDbId"
+                "seedLotTransactionDbId"
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Event.json:
```diff
@@ -95,7 +88,7 @@
         },
         "EventDateRange": {
             "properties": {
-                "rangeDbId": {
```
No range DbId in v.2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Event.json:
```diff
@@ -134,17 +127,17 @@
                 }
             },
             "required": [
-                "rangeDbId"
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Event.json:
```diff
@@ -134,17 +127,17 @@
-        "EventParameters": {
+        "EventParameter": {
             "properties": {
-                "parametersDbId": {
```
No parameterDbid in v2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Event.json:
```diff
@@ -208,7 +201,7 @@
             },
             "required": [
-                "parametersDbId"
+                "eventParameterDbId"
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/ObservationUnit.json:
```diff
@@ -183,7 +176,7 @@
         },
         "ObservationUnitPosition": {
             "properties": {
-                "positionDbId": {
```
no DbId in ObservationUnitPosition

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/ObservationUnit.json:
```diff
@@ -286,15 +281,15 @@
                 }
             },
             "required": [
-                "positionDbId"
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/OntologyReference.json:
```diff
@@ -37,9 +45,9 @@
-        "DocumentationLinks": {
+        "DocumentationLink": {
             "properties": {
-                "linkDbId": {
```
no DbId in v2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -101,9 +95,9 @@
-        "ValidValues": {
+        "ValidValue": {
             "properties": {
-                "validValuesDbId": {
```
not valid value DbId  in v2.1

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -143,17 +137,17 @@
             },
             "required": [
-                "validValuesDbId"
+                "validValueDbId"
```
see comment above

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -101,9 +95,9 @@
             }
         },
-        "ValidValues": {
+        "ValidValue": {
```
I actually think that ValidValues is correct over ValidValue in this case. The property would be validValues and it references 'one-to-one' an object of class 'ValidValues'

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -143,17 +137,17 @@
-        "Categories": {
+        "Category": {
             "properties": {
-                "categoriesDbId": {
```
There should be no DbId, for category, since it should be Value Type, and not an Entity.

daveneti commented on Specification/BrAPI-Schema/BrAPI-Phenotyping/Scale.json:
```diff
@@ -173,7 +167,7 @@
                 "validValues": {
                     "description": "associated valid values",
-                    "$ref": "../BrAPI-Phenotyping/Scale.json#/$defs/ValidValues",
+                    "$ref": "../BrAPI-Phenotyping/Scale.json#/$defs/ValidValue",
```
Should be ValidValues, see my comment above, sorry for my mistake in the previous review.

daveneti reviewed (changes requested):
@LzLang @BrapiCoordinatorSelby I finished the review. I think there is a little bit more work to do. Mainly because of the addition of DbIds in some 'Value Types', which changes the model. Also the GermplasmParent, Progeny/GermplasmChild needs some discussion.

daveneti commented:
@LzLang I finished the latest review. There are some minor changes that can be fixed easily and some more major questions that needs some input from @BrapiCoordinatorSelby that where in the original JSON Schema.

LzLang committed:
Changed schema according to the review

daveneti reviewed (changes requested):
@LzLang I think there is only two remaining issue. Sorry it has taken so long!

1. The issues are: Link from study to growth facility should be one-to-many, so the link should not be array. 
2. I think we can remove SourceGermplasm

I will make the changes in the branch directly, you can merge when you want. I will also run the validator.
